ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBackupManager.java

```diff
@@ -54,51 +64,113 @@ public CloudBackupManager(VerifiableProperties properties, CloudConfig cloudConf
+        // For CloudBackUpManger, at most one PartitionInfo in the list.
+        List<PartitionInfo> retList;
+        retList = (value == null ? new ArrayList<>() : value);
+        retList.add(partitionInfo);
+        return retList;
+      });
+    } else {
+      throw new ReplicationException("Failed to add partition " + partitionId + ", because no peer replicas found.");
+    }
+    // read replication token
+    for (RemoteReplicaInfo remoteReplicaInfo : remoteReplicaInfos) {
+      List<RemoteReplicaInfo.ReplicaTokenInfo> tokenInfos = persistor.retrieve(cloudReplica.getMountPath());
+      for (RemoteReplicaInfo.ReplicaTokenInfo tokenInfo : tokenInfos) {
+        DataNodeId dataNodeId = remoteReplicaInfo.getReplicaId().getDataNodeId();
+        if (dataNodeId.getHostname().equalsIgnoreCase(tokenInfo.getHostname())
+            && dataNodeId.getPort() == tokenInfo.getPort() && remoteReplicaInfo.getReplicaId()
+            .getReplicaPath()
+            .equals(tokenInfo.getReplicaPath())) {
+          logger.info("Read token for partition {} remote host {} port {} token {}", partitionId,
+              tokenInfo.getHostname(), tokenInfo.getPort(), tokenInfo.getReplicaToken());
+          remoteReplicaInfo.initializeTokens(tokenInfo.getReplicaToken());
+          remoteReplicaInfo.setTotalBytesReadFromLocalStore(tokenInfo.getTotalBytesReadFromLocalStore());
+          break;
         }
-        PartitionInfo partitionInfo = new PartitionInfo(remoteReplicas, partitionId, cloudStore, cloudReplica);
-        partitionsToReplicate.put(partitionId, partitionInfo);
-        partitionGroupedByMountPath.computeIfAbsent(cloudReplica.getMountPath(), key -> new ArrayList<>())
-            .add(partitionInfo);
-      } else {
-        throw new ReplicationException(
-            "Failed to back up partition " + partitionId + " because no peer replicas found.");
       }
     }
-    replicationMetrics.populatePerColoMetrics(numberOfReplicaThreads.keySet());
-    persistor =
-        new CloudTokenPersistor(replicaTokenFileName, partitionGroupedByMountPath, replicationMetrics, clusterMap,
-            factory, cloudDestination);
+    // create thread if necessary.
+    addRemoteReplicaInfoToReplicaThread(remoteReplicaInfos, true);
+    return true;
+  }
+
+  @Override
+  public void start() throws ReplicationException {
+    // Add listener for new coming assigned partition
+    virtualReplicatorCluster.addListener(new VirtualReplicatorClusterListener() {
+      @Override
+      public void onPartitionAdded(PartitionId partitionId) {
+        try {
+          addPartition(partitionId);
+        } catch (ReplicationException e) {
+          logger.error("Exception on adding partition: ", e);
+        }
+      }
+
+      @Override
+      public void onPartitionRemoved(PartitionId partitionId) {
+        removeRemoteReplicaInfoFromReplicaThread(partitionToPartitionInfo.get(partitionId).getRemoteReplicaInfos());
```

Comment:
When a partition is removed, we should flush its tokens so the new node hosting the partitions can get the latest offset.  I think we can call persistor.write(mountPath, true) here.